### PR TITLE
RxJava2 Autoconfiguration

### PR DESCRIPTION
**Pull request by lukaszguz**
Hi guys,
we developed auto configuration to RxJava2 based on RxJavaPlugins.
We have still had small problem because RxJavaPlugins' schedule handler is called when Observable, Single etc. are subscribe and when thread pool is changed.
In the first case we don't have problem but in the second RxJava's scheduler create worker and wrapped Runnable in own package scope wrapper so we don't know which type Runnable is.
The only thing we did is checked type using reflection.
http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Scheduler.html#scheduleDirect-java.lang.Runnable-

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@lukaszguz Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-sleuth&pullRequestId=775)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-sleuth&pullRequestId=775) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@lukaszguz Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-sleuth&pullRequestId=775)!

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+			}
+			Runnable wrappedAction = this.delegate != null ? this.delegate.apply(action)
+					: action;
+			return new SleuthRxJava2SchedulersHandler.TraceAction(this.tracer,
+					this.traceKeys, wrappedAction, this.threadsToSample);
+		}
+
+		private boolean isTraceActionDecoratedByRxWorker(Runnable action) {
+			try {
+				if (action instanceof Disposable) {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java]: this is likely an expensive check. maybe check w/ some rx maintainers to see if there's a better way?

**Comment by adriancole** [comment]
No problem.. awkward reflection is better than no progress. Keep us posted!

**Comment by lukaszguz** [comment]
I agree with you :) 
I asked on RxChannel and they got me very fast answer.
https://github.com/ReactiveX/RxJava/issues/5733

**Comment by lukaszguz** [comment]
I've opened PR in Rx
https://github.com/ReactiveX/RxJava/pull/5734

**Comment by adriancole** [comment]
>
> I've opened PR in Rx
> ReactiveX/RxJava#5734 <https://github.com/ReactiveX/RxJava/pull/5734>
>
hero status is on the way! I love it when things come together. Much
appreciated

**Comment by lukaszguz** [comment]
Hi again :)
RxJava2 had 2.1.8 release with all things needed to remove reflection from `ScheduleHandler`.

**Comment by marcingrzejszczak** [comment]
Great! BTW we've just started doing this - https://github.com/spring-cloud/spring-cloud-sleuth/issues/711 . That will mean that most of your work will be obsolete :/ We have to do this big refactoring in 2.0.x cause it's the last chance for us to break the compatibility. I'm not going to close this PR yet and ask you to redo it once we're done cause maybe I'll merge it later. I just wanted to keep you posted on this.

**Comment by lukaszguz** [comment]
"Life is life" :D, no problem I can wait and if you think that it will not be useful to you I have still tool to fix previous problem :)

**Comment by marcingrzejszczak** [comment]
We've merged the Brave instrumentation. Care to redo your work against the new master? :P

**Comment by lukaszguz** [comment]
I can try :D

**Comment by lukaszguz** [comment]
Ok, I think I've rewritten it on Brave implementation.

**Comment by marcingrzejszczak** [comment]
LGTM. @adriancole - maybe this should actually go to Brave? WDYT?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+					}
+					this.actual.run();
+					return;
+				}
+			}
+
+			Span span = this.parent;
+			boolean created = false;
+			if (span != null) {
+				span = this.tracer.joinSpan(this.parent.context());
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java]: not sure we want to use joinSpan like this. if you have a span, wouldn't you just use it?

**Comment by adriancole** [comment]
on the upstreaming, here's my 2p.

scheduler hook to propagate thread context should be decoupled from the auto-spanning code, as I'm sure the former is reusable, not quite sure the latter would be. Concerns I have are around the span policy and decoupling of things like which thread applies etc.

If starting with just rxjava2 trace propagation, we have a few advantages:
* there's existing need https://github.com/openzipkin/brave/issues/195
* it is smaller feature, so easier to acclimate to the different review and code style culture
* it takes out the polarizing code which helps also with ^^

Dare to give it a go? If so, it would be under brave's repo instrumentation/rxjava2 and start strictly with propagation, covering the scenarios where a user or another tool can create spans that continue from whatever was the ancestor. If you go for it, avoid any dependencies rxjava2 doesn't have, and any features that aren't strictly required for propagation to work. For example, when not auto-spanning, you can safely propagate trace context regardless I think.

If you go that route, open a PR, citiing brave's issue 195 cc'ing people from that issue and also the rxjava person you were able to contact formerly. Once that's settled, we'd talk about auto-spanning things

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java (hunk 2 above)
Comment (lukaszguz) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java]: The documentation contains "Joining is re-using the same trace and span ids extracted from an incoming RPC request. This should not be used for messaging operations, as nextSpan(TraceContextOrSamplingFlags) is a better choice." of course it doesn't have to be RPC call but observer or subscriber or whatever call. SleuthRxJavaSchedulersHook.TraceAction also has the same logic (join span).
What do @adriancole think about it?

**Comment by lukaszguz** [comment]
I do not know if I understood correctly. You mean remove RxJava2Autoconfiguration from sleuth-core and add new brave's repo instrumentation/rxjava2 which comprehensively solve tracing in rx2?

**Comment by adriancole** [comment]
I think the key here is that it isn't remote as you literally have a
(local) reference to the span you want. For example you didnt extract the
span from headers, you dont need to double check sampling status etc. Join
is a remote special case for RPC. Definitely we can run out join as
appropriate.

More important than documentation nuance is this.. why are we taking a span
we have and the turning it back into a span. What happens if you were to
use the reference (to what is named parent) directly? I think you can.

There is probably going to be a few places where the api mapping from
sleuth to brave is slightly different. This is probably one of those
things. In brave joinSpan is only for remote. If you have a span locally
just use it. Simple as that. If you want to make a child, what you had
(nextSpan) is correct.

On 22 Jan 2018 9:27 pm, "lukaszguz" <notifications@github.com> wrote:

*@lukaszguz* commented on this pull request.
------------------------------

In spring-cloud-sleuth-core/src/main/java/org/springframework/c
loud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java
<https://github.com/spring-cloud/spring-cloud-sleuth/pull/775#discussion_r162935204>
:

> +			for (String threadToIgnore : this.threadsToIgnore) {
+				if (threadName.matches(threadToIgnore)) {
+					if (log.isTraceEnabled()) {
+						log.trace(String.format(
+								"Thread with name [%s] matches the regex [%s]. A span will
not be created for this Thread.",
+								threadName, threadToIgnore));
+					}
+					this.actual.run();
+					return;
+				}
+			}
+
+			Span span = this.parent;
+			boolean created = false;
+			if (span != null) {
+				span = this.tracer.joinSpan(this.parent.context());

The documentation contains "Joining is re-using the same trace and span ids
extracted from an incoming RPC request. This should not be used for
messaging operations, as nextSpan(TraceContextOrSamplingFlags) is a better
choice." of course it doesn't have to be RPC call but observer or
subscriber or whatever call. SleuthRxJavaSchedulersHook.TraceAction also
has the same logic (join span).
What do @adriancole <https://github.com/adriancole> think about it?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/spring-cloud/spring-cloud-sleuth/pull/775#discussion_r162935204>,
or mute the thread

**Comment by adriancole** [comment]
Think of sleuth as focused on autoconfiguration. If you look at bravd there
are many instrumentation there some we havent autoconfigured. So apart from
the autoconfiguration part, rebuild rxjava2 in brave in a composable way,
starting with just propagation. This code would have no spring version
binding so can work in or out of spring or any version of spring. Sleuth
would add the smart defaults maybe compose things like thread conventions
etc.

Have a look at existing brave and code here that binds it. Iirc sleuth
reuses brave apach http client stuff so maybe a good thing to look at.

So yes, build comprehensive there except spring part.... and start small..
just propagation as many wont want to make spans for every transition.

**Comment by lukaszguz** [comment]
Ok, I think I've got it. Your explanation about "joins" and "brave" is enough for me :) so I will remove "joins" from autoconfig.
For now I don't know if I will have enough time to brave's development but I will see.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			RxJavaPlugins.reset();
+			RxJavaPlugins.setScheduleHandler(
+					new ScheduleHandler(tracer, traceKeys, threadsToSample, delegate));
+		}
+		catch (Exception e) {
+			log.error("Failed to register Sleuth RxJava SchedulersHook", e);
+		}
+	}
+
+	private void logCurrentStateOfRxJavaPlugins() {
```
[the remark this answers is not in the document]
Comment (lukaszguz) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java]: Changed.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java (hunk 2 above)
Comment (lukaszguz) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java]: Join is removed.

**Comment by lukaszguz** [comment]
Could @adriancole look at https://github.com/openzipkin/brave/compare/master...lukaszguz:master ?
And say what you think about it?

**Comment by adriancole** [comment]
Yep will do later today or tomorrow. Shaking off jetlag

**Comment by lukaszguz** [comment]
Sure :) I wish you health ;)

**Comment by adriancole** [comment]
opened https://github.com/openzipkin/brave/issues/664 for a "continue-only" scheduler hook for rxjava2. This would not trace rxjava2, but it could prevent broken traces. Once this is out, we should revisit this (which requires trace propagation, but does more as it is also automatically creating spans and configuring sleuth)

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	SleuthRxJava2SchedulersHandler(Tracer tracer, TraceKeys traceKeys,
+			List<String> threadsToSample) {
+		try {
+			Function<? super Runnable, ? extends Runnable> delegate = RxJavaPlugins
+					.getScheduleHandler();
+			if (delegate instanceof SleuthRxJava2SchedulersHandler.ScheduleHandler) {
+				return;
+			}
+			logRegisterRxJavaPlugins();
+			RxJavaPlugins.reset();
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java]: just reviewing this closer while pulling out current trace context code.

ps this resets all plugins to null/default.. is this what you wanted to do? What if other plugins besides scheduler were present?

**Comment by marcingrzejszczak** [comment]
Also, I wonder if the current testing suite is enough. Does the instrumentation work with flat maps and more sophisticated operators? Our initial reactor one didn't :|

**Comment by adriancole** [comment]
this is a big job :P

**Comment by lukaszguz** [comment]
Sorry for silent I was on holiday :)

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java (hunk 4 above)
Comment (lukaszguz) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/rxjava2/SleuthRxJava2SchedulersHandler.java]: I little based on plugin to RxJava1 therefore I left this line.

**Comment by marcingrzejszczak** [comment]
What do we do about this? :P

**Comment by adriancole** [comment]
dunno we can put it on the hopper for discussion next week. right now we
have rxjava2 propagation only https://etherpad.net/p/zipkin_fukuoka

On Wed, Aug 1, 2018 at 9:02 PM, Marcin Grzejszczak <notifications@github.com

**Comment by marcingrzejszczak** [comment]
Let me close this for now. I guess we would be adding this in Brave anyways.

**Pull request closed without merging**